src/main/groovy/info/solidsoft/gradle/pitest/AggregateReportTask.groovy

```diff
@@ -0,0 +1,181 @@
+        mutationFiles.from = collectMutationFiles(projectsWithPitest)
+
+        lineCoverageFiles = project.objects.fileCollection()
+        lineCoverageFiles.from = collectLineCoverageFiles(projectsWithPitest)
+    }
+
+    @TaskAction
+    void aggregate() {
+        logger.info("Aggregating pitest reports")
+
+        ReportAggregator.Builder builder = ReportAggregator.builder()
+
+        lineCoverageFiles.each { file -> builder.addLineCoverageFile(file) }
+        mutationFiles.each { file -> builder.addMutationResultsFile(file) }
+        sourceDirs.each { file -> builder.addSourceCodeDirectory(file) }
+        additionalClasspath.each { file -> builder.addCompiledCodeDirectory(file) }
+
+        ReportAggregator aggregator = builder.resultOutputStrategy(new DirectoryResultOutputStrategy(
+            reportDir.asFile.get().absolutePath,
+            new UndatedReportDirCreationStrategy()))
+            .build()
+        aggregator.aggregateReport()
+
+        logger.info("Aggregated report ${reportFile.asFile.get().absolutePath}")
+    }
+
+    private static Set<Provider<File>> collectMutationFiles(Set<Project> pitestProjects) {
+        return pitestProjects.stream()
+            .map { prj -> prj.extensions.getByType(PitestPluginExtension) }
+            .map { extension -> extension.reportDir.file(MUTATION_FILE_NAME) }
+            .collect(Collectors.toSet())
+    }
+
+    private static Set<Provider<File>> collectLineCoverageFiles(Set<Project> pitestProjects) {
+        return pitestProjects.stream()
+            .map { prj -> prj.extensions.getByType(PitestPluginExtension) }
+            .map { extension -> extension.reportDir.file(LINE_COVERAGE_FILE_NAME) }
+            .collect(Collectors.toSet())
+    }
+
+    DirectoryProperty getReportDir() {
+        return reportDir
+    }
+
+    ConfigurableFileCollection getAdditionalClasspath() {
+        return additionalClasspath
+    }
+
+    ConfigurableFileCollection getMutationFiles() {
+        return mutationFiles
+    }
+
+    ConfigurableFileCollection getLineCoverageFiles() {
+        return lineCoverageFiles
+    }
+
+    ConfigurableFileCollection getSourceDirs() {
+        return sourceDirs
+    }
+
+    private File getReportBaseDirectory() {
+        // if Java plugin configured on root project
+        if (project.extensions.findByType(ReportingExtension)) {
+            return project.extensions.getByType(ReportingExtension).baseDir
+        }
+        return new File(project.buildDir, "reports")
+    }
+
+    private Set<Project> getProjectsWithPitestPlugin() {
+        Set<Project> projects = [] as Set
+        if (isRootPitestConfigured()) {
+            projects.add(project)
+        }
+        projects.addAll(getSubprojectsWithPitest())
+        return projects
+    }
+
+    private boolean isRootPitestConfigured() {
+        return project.plugins.hasPlugin(PitestPlugin.PLUGIN_ID) && project.extensions.findByType(PitestPluginExtension)
```

Comment:
I'm back to that topic and it looks quite promising. However, adjusting the code, I wonder if there is a reason why you for the root project check for both plugin and its extension, but for subproject you only check for the plugin existence?

Extension should be applied automatically when the plugin is applied, so maybe it is superfluous?

(please do not commit any changes, I'm working on dependencies which I plan to push to that branch once ready)

Reply:
`Extension should be applied automatically when the plugin is applied, so maybe it is superfluous?`

You are completely right.